lib/Conversion/AffineToPipeline/AffineToPipeline.cpp

```diff
@@ -397,48 +431,101 @@ LogicalResult AffineToPipeline::createPipelinePipeline(
 
   // Maintain mappings of values in the loop body and results of stages,
   // initially populated with the iter args.
-  BlockAndValueMapping valueMap;
-  for (size_t i = 0; i < iterArgs.size(); ++i)
-    valueMap.map(forOp.getBody()->getArgument(i),
-                 pipeline.getStagesBlock().getArgument(i));
+  BlockAndValueMapping iterValueMap;
+  for (size_t i = 0; i < forOp.getBody()->getNumArguments(); ++i)
+    iterValueMap.map(forOp.getBody()->getArgument(i),
+                     pipeline.getStagesBlock().getArgument(i));
 
   // Create the stages.
   Block &stagesBlock = pipeline.getStagesBlock();
   builder.setInsertionPointToStart(&stagesBlock);
 
   // Iterate in order of the start times.
   SmallVector<unsigned> startTimes;
-  for (auto group : startGroups)
+  for (const auto &group : startGroups)
     startTimes.push_back(group.first);
   llvm::sort(startTimes);
 
   DominanceInfo dom(getOperation());
+  SmallVector<DenseSet<Value>> registerValues;
+  SmallVector<SmallVector<mlir::Type>> registerTypes;
+  SmallVector<BlockAndValueMapping> valueMaps;
+  DenseMap<Operation *, std::pair<unsigned, unsigned>> pipeTimes;
```

Comment:
I know I haven't been great about this, but please add comments explaining what all these new data structures are for. Especially for the pair of numbers: I see that it represents a start/end time step pair, but it would be good to document this in a comment.